catalog/core/catalog-core-api-impl/src/main/java/ddf/catalog/util/impl/SourcePoller.java

```diff
@@ -13,110 +13,285 @@
+  private static final Logger LOGGER = LoggerFactory.getLogger(SourcePoller.class);
 
-  private int interval = 1;
+  private static final long MINIMUM_TIMEOUT = 1;
 
-  private ScheduledExecutorService scheduler;
+  private final ExecutorService pollThreadPool;
 
-  private ScheduledFuture<?> handle;
+  private final ExecutorService pollTimeoutWatcherThreadPool;
 
-  private SourcePollerRunner runner;
+  private final Cache<SourceKey, T> cache = CacheBuilder.newBuilder().maximumSize(1000).build();
 
   /**
-   * Constructor to schedule the SourcePollerRunner to execute immediately and at a fixed interval,
-   * currently set at every 1 minute. This constructor is invoked by the CatalogFramework's
-   * blueprint.
-   *
-   * @param incomingRunner the SourcePollerRunner to use for polling
+   * @throws NullPointerException if any of the parameters are {@code null}
+   * @throws IllegalArgumentException if {@code valueName} is empty
+   * @throws IllegalArgumentException if the {@code pollThreadPool} or {@code
+   *     pollTimeoutWatcherThreadPool} {@link ExecutorService#isShutdown()}
    */
-  public SourcePoller(SourcePollerRunner incomingRunner) {
-
-    this.runner = incomingRunner;
-
-    scheduler =
-        Executors.newSingleThreadScheduledExecutor(
-            StandardThreadFactoryBuilder.newThreadFactory("sourcePollerThread"));
+  public SourcePoller(
+      final ExecutorService pollThreadPool, final ExecutorService pollTimeoutWatcherThreadPool) {
+    this.pollThreadPool = notShutdown(notNull(pollThreadPool));
+    this.pollTimeoutWatcherThreadPool = notShutdown(notNull(pollTimeoutWatcherThreadPool));
+  }
 
-    handle = scheduler.scheduleAtFixedRate(runner, INITIAL_DELAY, interval, TimeUnit.MINUTES);
+  public void destroy() {
+    MoreExecutors.shutdownAndAwaitTermination(pollThreadPool, 5, TimeUnit.SECONDS);
+    MoreExecutors.shutdownAndAwaitTermination(pollTimeoutWatcherThreadPool, 5, TimeUnit.SECONDS);
   }
 
   /**
-   * Retrieves a {@link CachedSource} which contains cached values from the specified {@link
-   * Source}. Returns a {@link Source} with values from the last polling interval. If the {@link
-   * Source} is not known, null is returned.
+   * This method does not block.
    *
-   * @param source the source to get the {@link CachedSource} for
-   * @return a {@link CachedSource} which contains cached values
+   * @return TODO
+   * @throws NullPointerException if the {@code source} is {@code null}
    */
-  public CachedSource getCachedSource(Source source) {
-    return runner.getCachedSource(source);
+  @SuppressWarnings(
+      "squid:S2789" /*The cache contains Optional<T> values. The Optional<T> returned by cache.getIfPresent is checked for null to see if the cache has an entry for the key*/)
+  public Optional<T> getCachedValue(final Source source) {
+    final SourceKey sourceKey = new SourceKey(source);
+
+    final T cachedValue = cache.getIfPresent(sourceKey);
```

Comment:
```suggestion
    return Optional.ofNullable(cache.getIfPresent(sourceKey));
```

Reply:
I want to log the cached value if it is present, so I would need to get the value form the optional that was just created, so I think this improvement makes the method a little less readable:
```java
final Optional<V> optionalCachedValue = Optional.ofNullable(cache.getIfPresent(notNull(key)));
if (!optionalCachedValue.isPresent()) {
  LOGGER.debug(
      "{} is unknown. This only happens for keys that have not yet successfully completed a poll yet. Returning \"unknown\"",
      key);
  return optionalCachedValue;
} else {
  LOGGER.trace("The cached value for {} is {}", key, optionalCachedValue.get());
  return optionalCachedValue;
}
```